From 5079287065c74194e3e068e7962bedc02ccac4fd Mon Sep 17 00:00:00 2001
From: Omar Eissa <oeissa@google.com>
Date: Wed, 27 Aug 2025 13:55:24 +0000
Subject: [PATCH] Don't allow 3p apps to become voice recognizer automatically

Currently, If there is no default recognizer set using config_systemSpeechRecognizer, any 3p app defining recongition service could become the default recognizer automatically without user selection. Make sure that we only automatically select system apps rather random 3p app.

Test: Manual, install 3p app defining recognizer when no default or
system recognizer is available on the device
Bug: 352518318
Flag: EXEMPT CVE_FIX
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:0bc6ced90e2ee1e6140a4d87ee62923bdfd2ca93)
Merged-In: I52681da7381be4ebe6cf20b639508311f3a90eba
Change-Id: I52681da7381be4ebe6cf20b639508311f3a90eba
---
 .../VoiceInteractionManagerService.java          | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/services/voiceinteraction/java/com/android/server/voiceinteraction/VoiceInteractionManagerService.java b/services/voiceinteraction/java/com/android/server/voiceinteraction/VoiceInteractionManagerService.java
index 605af03f112f..33a4392d3562 100644
--- a/services/voiceinteraction/java/com/android/server/voiceinteraction/VoiceInteractionManagerService.java
+++ b/services/voiceinteraction/java/com/android/server/voiceinteraction/VoiceInteractionManagerService.java
@@ -852,12 +852,20 @@ public class VoiceInteractionManagerService extends SystemService {
                         }
                     }
                 }
-                if (numAvailable > 1) {
-                    Slog.w(TAG, "more than one voice recognition service found, picking first");
+
+                // If prefPackage isn't found then only default to system recognizer.
+                // prefPackage could be either the current recognizer or the default recognizer.
+                for (int i = 0; i < numAvailable; i++) {
+                    ServiceInfo serviceInfo = available.get(i).getServiceInfo();
+                    if ((serviceInfo.applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) == 0) {
+                        continue;
+                    }
+                    return new ComponentName(serviceInfo.packageName, serviceInfo.name);
                 }
 
-                ServiceInfo serviceInfo = available.get(0).getServiceInfo();
-                return new ComponentName(serviceInfo.packageName, serviceInfo.name);
+                Slog.w(TAG, "no auto selectable voice recognition services found for user "
+                        + userHandle);
+                return null;
             }
         }
 
-- 
2.52.0.rc1.455.g30608eb744-goog

